plugins/openstack/src/main/java/org/eclipse/xpanse/orchestrator/plugin/openstack/monitor/keystone/CustomEndPointResolver.java

```diff
@@ -0,0 +1,250 @@
+        if (p.type == null) {
+            return p.access.getEndpoint();
+        }
+
+        Key
+                key = Key.of(p.access.getCacheIdentifier(), p.type, p.perspective, p.region);
+        String url = CACHE.get(key);
+
+        if (url != null) {
+            return url;
+        }
+
+        url = resolveV2(p);
+
+        if (url != null) {
+            return url;
+        } else if (p.region != null) {
+            throw RegionEndpointNotFoundException.create(p.region, p.type.getServiceName());
+        }
+
+        return p.access.getEndpoint();
+    }
+
+    @Override
+    public String findURLV3(URLResolverParams p) {
+
+        if (p.type == null) {
+            return p.token.getEndpoint();
+        }
+
+        Key
+                key = Key.of(p.token.getCacheIdentifier(), p.type, p.perspective, p.region);
+
+        String url = CACHE.get(key);
+
+        if (url != null) {
+            return url;
+        }
+
+        url = resolveV3(p);
+
+        if (url != null) {
+            CACHE.put(key, url);
+            return url;
+        } else if (p.region != null) {
+            throw RegionEndpointNotFoundException.create(p.region, p.type.getServiceName());
+        }
+
+        return p.token.getEndpoint();
+    }
+
+    private String resolveV2(URLResolverParams p) {
+        SortedSetMultimap<String, ? extends Access.Service> catalog =
+                p.access.getAggregatedCatalog();
+        SortedSet<? extends Access.Service> services = catalog.get(p.type.getServiceName());
+
+        if (services.isEmpty()) {
+            services = catalog.get(p.type.getType());
+        }
+
+        if (!services.isEmpty()) {
+            Access.Service sc = p.getV2Resolver().resolveV2(p.type, services);
+            for (Endpoint ep : sc.getEndpoints()) {
+                if (p.region != null && !p.region.equalsIgnoreCase(ep.getRegion())) {
+                    continue;
+                }
+
+                if (sc.getServiceType() == ServiceType.NETWORK) {
+                    sc.getEndpoints().get(0).toBuilder().type(sc.getServiceType().name());
+                }
+
+                if (p.perspective == null) {
+                    return getEndpointUrl(p.access, ep);
+                }
+
+                return switch (p.perspective) {
+                    case ADMIN -> ep.getAdminURL().toString();
+                    case INTERNAL -> ep.getInternalURL().toString();
+                    default -> ep.getPublicURL().toString();
+                };
+            }
+        } else {
+            //if no catalog returned, if is identity service, just return endpoint
+            if (ServiceType.IDENTITY.equals(p.type)) {
+                return p.access.getEndpoint();
+            }
+        }
+        return null;
+    }
+
+    private String resolveV3(URLResolverParams urlResolverParams) {
+        Token token = urlResolverParams.token;
+
+        //in v3 api, if user has no default project, and token is unscoped,
```

Comment:
```suggestion
        // in v3 api, if user has no default project, and token is unscoped,
```
or, can we do it with `/**/` here?